src/passes/GlobalStructInference.cpp

```diff
@@ -187,46 +187,113 @@ struct GlobalStructInference : public Pass {
+        // single comparison, add tradeoffs that may not be worth it, and a
+        // single value (or no value) is already handled by other passes.
+        //
+        // That situation may involve more than two globals. For example we may
+        // have three relevant globals, but two may have the same value. In that
+        // case we can compare against the third:
+        //
+        //  $global0: (struct.new $Type (i32.const 42))
+        //  $global1: (struct.new $Type (i32.const 42))
+        //  $global2: (struct.new $Type (i32.const 1337))
+        //
+        // (struct.get $Type (ref))
+        //   =>
+        // (select
+        //   (i32.const 1337)
+        //   (i32.const 42)
+        //   (ref.eq (ref) $global2))
+        auto& globals = iter->second;
+        if (globals.size() < 2) {
+          return;
+        }
+
+        // Find the constant values and which globals correspond to them.
+        // TODO: SmallVectors?
         std::vector<Literal> values;
+        std::vector<std::vector<Name>> globalsForValue;
+
+        // Check if the relevant fields contain constants.
+        auto& wasm = *getModule();
+        auto fieldType = field.type;
         for (Index i = 0; i < globals.size(); i++) {
-          auto* structNew = wasm.getGlobal(globals[i])->init->cast<StructNew>();
+          Name global = globals[i];
+          auto* structNew = wasm.getGlobal(global)->init->cast<StructNew>();
+          Literal value;
           if (structNew->isWithDefault()) {
-            values.push_back(Literal::makeZero(fieldType));
+            value = Literal::makeZero(fieldType);
           } else {
             auto* init = structNew->operands[fieldIndex];
             if (!Properties::isConstantExpression(init)) {
               // Non-constant; give up entirely.
               return;
             }
-            values.push_back(Properties::getLiteral(init));
+            value = Properties::getLiteral(init);
+          }
+
+          // Process the current value, comparing it against the previous.
+          auto found = std::find(values.begin(), values.end(), value);
+          if (found == values.end()) {
+            // This is a new value.
+            assert(values.size() <= 2);
+            if (values.size() == 2) {
+              // Adding this value would mean we have too many, so give up.
+              return;
+            }
+            values.push_back(value);
+            globalsForValue.push_back({global});
+          } else {
+            // This is an existing value.
+            Index index = found - values.begin();
+            globalsForValue[index].push_back(global);
           }
         }
 
+        // We have some globals (at least 2), and so must have at least one
+        // value.
+        assert(values.size() >= 1);
+        if (values.size() != 2) {
+          // More than two values to pick from, so we'd need more than one
+          // comparison. Give up.
+          return;
+        }
```

Comment:
This is already handled up on line 249, isn't it? Maybe this is necessary to catch `values.size() == 1`, in which case the condition and comment should be updated. Does another pass handle the case where there are multiple globals but just one value? If not, maybe we should handle that here.

Reply:
Updating the comments and conditions, good point.

Yes, CFP should handle the case of multiple globals and one value, I'll mention that.